File: Binary_Search/LC_981_Time_Based_Key_Value_Store.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.22

class TimeMap {
public:
    TimeMap() {
        
    }
    
    void set(string key, string value, int timestamp) {
        mapping[key].push_back({timestamp, value});
    }
    
    string get(string key, int timestamp) {
        if (!mapping.count(key)){
            return "";
        }
        int idx = find_idx(mapping[key], timestamp);
        if (idx == -1){
            return "";
        }
        return mapping[key][idx].second;
    }
private:
    unordered_map<string, vector<pair<int, string>>> mapping;
    // upper bound function
    int find_idx(vector<pair<int, string>>& nums, int val){
        int l = 0, r = nums.size();
        while (l < r){
            int mid = l + (r-l)/2;
            int t = nums[mid].first;
            if (t > val){
                r = mid;
            }
            else{
                l = mid + 1;
            }
        }
        return l - 1;
    }
};

File: Binary_Search/LC_981_Time_Based_Key_Value_Store.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.22

class TimeMap:

    def __init__(self):
        self.map = collections.defaultdict(list)
    def set(self, key: str, value: str, timestamp: int) -> None:
        self.map[key].append((timestamp, value))

    def get(self, key: str, timestamp: int) -> str:
        content = self.map[key]
        #print(content)
        l, r = 0, len(content)
        while l < r:
            mid = l + (r-l)//2
            if content[mid][0] > timestamp:
                r = mid
            else:
                l = mid + 1
        if l-1 < 0:
            return ""
        return content[l-1][1]